### PR TITLE
- fixed inserting generated primary key

### PR DESCRIPTION
**Pull request by bogensberger**
- fixed routing values on generated columns

### CODE REVIEW

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java [hunk 1]
```diff
@@ -903,6 +903,29 @@ public void testInsertGeneratedPrimaryKeyValue() throws Exception {
...
+
+    @Test
+    public void testInsertGeneratedPrimaryKeyPartMissing() throws Exception {
+        execute("create table test(col1 integer, " +
+                "                  col2 integer, " +
+                "                  col3 as col1 + 1 primary key, " +
+                "                  col4 as col2 + 1 primary key)");
+        ensureYellow();
+        expectedException.expect(SQLActionException.class);
+        expectedException.expectMessage("Missing primary key values");
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: should be an analyzer test

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java [hunk 2]
```diff
@@ -903,6 +903,29 @@ public void testInsertGeneratedPrimaryKeyValue() throws Exception {
     }
 
     @Test
+    public void testInsertGeneratedPrimaryKeyMissing() throws Exception {
+        execute("create table test(col1 as col2 + 1 primary key, " +
+                "                  col2 integer, " +
+                "                  col3 integer)");
+        ensureYellow();
+        expectedException.expect(SQLActionException.class);
+        expectedException.expectMessage("Primary key is required but is missing from the insert statement");
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: should be an analyzer test

File: sql/src/main/java/io/crate/analyze/AbstractInsertAnalyzer.java [hunk 3]
```diff
@@ -83,12 +84,35 @@ protected void handleInsertColumns(Insert node, int maxInsertValues, AbstractIns
...
+                throw new IllegalArgumentException("Clustered by value is required but is missing from the insert statement");
+            }
+        }
+    }
+
+    private boolean checkReferencesForGeneratedColumn(ColumnIdent columnIdent, AbstractInsertAnalyzedStatement context) {
+        ReferenceInfo referenceInfo = context.tableInfo().getReferenceInfo(columnIdent);
+        if (referenceInfo instanceof GeneratedReferenceInfo) {
+            for (ReferenceInfo ref : ((GeneratedReferenceInfo) referenceInfo).referencedReferenceInfos()) {
+                if (context.containsReferenceInfo(ref)) {
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/AbstractInsertAnalyzer.java]: isChild must be checked also

File: sql/src/main/java/io/crate/analyze/AbstractInsertAnalyzer.java [hunk 4]
```diff
@@ -83,12 +84,35 @@ protected void handleInsertColumns(Insert node, int maxInsertValues, AbstractIns
 
         if (context.primaryKeyColumnIndices().size() == 0 &&
                 !(context.tableInfo().primaryKey().isEmpty() || context.tableInfo().hasAutoGeneratedPrimaryKey())) {
-            throw new IllegalArgumentException("Primary key is required but is missing from the insert statement");
+            boolean referenced = false;
+            for (ColumnIdent columnIdent : context.tableInfo().primaryKey()) {
+                if (checkReferencesForGeneratedColumn(columnIdent, context)) {
+                    referenced= true;
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/AbstractInsertAnalyzer.java]: missing space

File: sql/src/test/java/io/crate/analyze/InsertFromValuesAnalyzerTest.java [hunk 5]
```diff
@@ -1051,4 +1077,18 @@ public void testInsertMultipleValuesWithGeneratedColumnAndTooFewValuesInSecondVa
...
+        expectedException.expectMessage("Primary key is required but is missing from the insert statement");
+        analyze("INSERT INTO generated_pk_column (color) values ('green')");
+    }
+
+    @Test
+    public void testGeneratedKeyPrimaryKeyPartMissing() throws Exception {
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("Missing primary key values");
+        analyze("INSERT INTO generated_pk_column (serial_no) values (1)");
+    }
```
Comment (seut) [inline sql/src/test/java/io/crate/analyze/InsertFromValuesAnalyzerTest.java]: pls add regression test for missing routing

Comment (seut) [inline sql/src/test/java/io/crate/analyze/InsertFromValuesAnalyzerTest.java]: also please add a regression test using generated column with subscript (see my comment about isChild check above)

File: sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java [hunk 6]
```diff
@@ -927,4 +927,29 @@ public void testInsertGeneratedPrimaryKeyValueGiven() throws Exception {
         expectedException.expectMessage("Given value 0 for generated column does not match defined generated expression value 4");
         execute("insert into test(col1, col2) values (1, 0)");
     }
+
+    @Test
+    public void testInsertFromSubQueryMissingPrimaryKeyValues() throws Exception {
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: if possible, moving this into a analyzer test would be nice

Comment (bogensberger) [inline sql/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java]: can't move that, it's evaluated the RowShardResolver which is used by the ColumnIndexWriterProjector

**Comment by seut** [comment]
ack

**Pull request merged**